deployment/terraform/keycloak.go

```diff
@@ -0,0 +1,205 @@
+	if err != nil {
+		return fmt.Errorf("failed to upload keycloak service file: %w", err)
+	}
+
+	// Enable and start service
+	_, err = sshc.RunCommand("sudo systemctl enable keycloak")
+	if err != nil {
+		return fmt.Errorf("failed to enable keycloak service: %w", err)
+	}
+
+	// Using restart to apply any possible changes to the service
+	_, err = sshc.RunCommand("sudo systemctl restart keycloak")
+	if err != nil {
+		return fmt.Errorf("failed to restart keycloak service: %w", err)
+	}
+
+	// Wait for keycloak to start
+	url := fmt.Sprintf("http://%s:8080/health", t.output.KeycloakServer.PublicDNS)
+	timeout := time.After(120 * time.Second) // yes, is **that** slow
+	for {
+		resp, err := http.Get(url)
+		if err == nil && resp.StatusCode == http.StatusOK {
+			break
+		}
+		mlog.Info("Service not up yet, waiting...")
+		select {
+		case <-timeout:
+			return errors.New("timeout: keycloak service is not responding")
+		case <-time.After(5 * time.Second):
+		}
+	}
+
+	_, err = sshc.RunCommand(fmt.Sprintf("%s/kcadm.sh config credentials --server http://127.0.0.1:8080 --user %s --password %s --realm master", keycloakBinPath, t.config.ExternalAuthProviderSettings.KeycloakAdminUser, t.config.ExternalAuthProviderSettings.KeycloakAdminPassword))
+	if err != nil {
+		return fmt.Errorf("failed to authenticate keycload admin: %w", err)
+	}
+
+	// Disable SSL requirement on master realm to allow http connections on the web interface
+	_, err = sshc.RunCommand(keycloakBinPath + "/kcadm.sh update realms/master -s sslRequired=NONE")
+	if err != nil {
+		return fmt.Errorf("failed to disable ssl requirement: %w", err)
+	}
+
+	// Disable SSL requirement on the realm set up by the loadtest
+	if t.config.ExternalAuthProviderSettings.KeycloakRealmName != "" && t.config.ExternalAuthProviderSettings.KeycloakRealmName != "master" {
+		_, err = sshc.RunCommand(fmt.Sprintf(keycloakBinPath+"/kcadm.sh update realms/%s -s sslRequired=NONE", t.config.ExternalAuthProviderSettings.KeycloakRealmName))
+		if err != nil {
+			return fmt.Errorf("failed to disable ssl requirement: %w", err)
+		}
+	}
+
+	// Populate users
+	if t.config.ExternalAuthProviderSettings.GenerateUsersCount > 0 {
+		if err := t.populateKeycloakUsers(sshc); err != nil {
+			return fmt.Errorf("failed to populate keycloak users: %w", err)
+		}
+
+		mlog.Info("Overriding the users file path with the generated one from keycloak")
+		t.config.UsersFilePath = t.getAsset("keycloak-users.txt")
+	}
+
+	mlog.Info("Keycloak configured")
+
+	return nil
+}
+
+func (t *Terraform) populateKeycloakUsers(sshc *ssh.Client) error {
+	keycloakBinPath := "/opt/keycloak/keycloak-" + t.config.ExternalAuthProviderSettings.KeycloakVersion + "/bin"
+	usersTxtPath := t.getAsset("keycloak-users.txt")
+
+	// Check if users file exists. Prevents from creating users multiple times.
+	if _, err := os.Stat(usersTxtPath); err == nil || os.IsExist(err) {
+		return nil
+	}
```

Comment:
Not sure if this is a dumb question, but: could it happen that the users file exists but that it doesn't contain the users we want? For example, we create a deployment with 10 users and then recreate it with 50.

Reply:
Yeah, I see that hapenning if that's an usual way of using the load test. Alternatively you could remove the file and the keycloak database to start from scratch.

Reply:
+1 to this. We should do a quick check to ensure that # of users in keycloak file is equal to # users in users.txt file. And if possible, have a way to add new users, or fail clearly in the logs. But importantly, there should never be a case of some users partially logging in via keycloak and some via email/pass.

Reply:
Changed it to create the number of users setup in the configuration. If the number is higher it will create more users accordingly, using the number of lines in the `users.txt` file as the source. It will skip user creation if the number of lines is the same or higher than the expected number of users.